source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -73,89 +205,161 @@ void Filter::onReadWorker() {
       throw EnvoyException("failed to read proxy protocol");
     }
 
-    Network::Address::IpVersion protocol_version;
-    Network::Address::InstanceConstSharedPtr remote_address;
-    Network::Address::InstanceConstSharedPtr local_address;
-
     // TODO(gsagula): parseInternetAddressAndPort() could be modified to take two string_view
     // arguments, so we can eliminate allocation here.
     if (line_parts[1] == "TCP4") {
-      protocol_version = Network::Address::IpVersion::v4;
-      remote_address = Network::Utility::parseInternetAddressAndPort(
-          std::string{line_parts[2]} + ":" + std::string{line_parts[4]});
-      local_address = Network::Utility::parseInternetAddressAndPort(
-          std::string{line_parts[3]} + ":" + std::string{line_parts[5]});
+      proxy_protocol_header_.emplace(
+          WireHeader{0, Network::Address::IpVersion::v4,
+                     Network::Utility::parseInternetAddressAndPort(
+                         std::string{line_parts[2]} + ":" + std::string{line_parts[4]}),
+                     Network::Utility::parseInternetAddressAndPort(
+                         std::string{line_parts[3]} + ":" + std::string{line_parts[5]})});
     } else if (line_parts[1] == "TCP6") {
-      protocol_version = Network::Address::IpVersion::v6;
-      remote_address = Network::Utility::parseInternetAddressAndPort(
-          "[" + std::string{line_parts[2]} + "]:" + std::string{line_parts[4]});
-      local_address = Network::Utility::parseInternetAddressAndPort(
-          "[" + std::string{line_parts[3]} + "]:" + std::string{line_parts[5]});
+      proxy_protocol_header_.emplace(
+          WireHeader{0, Network::Address::IpVersion::v6,
+                     Network::Utility::parseInternetAddressAndPort(
+                         "[" + std::string{line_parts[2]} + "]:" + std::string{line_parts[4]}),
+                     Network::Utility::parseInternetAddressAndPort(
+                         "[" + std::string{line_parts[3]} + "]:" + std::string{line_parts[5]})});
     } else {
       throw EnvoyException("failed to read proxy protocol");
     }
+  }
+}
 
-    // Error check the source and destination fields. Most errors are caught by the address
-    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
-    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
-    // address.
-    const auto remote_version = remote_address->ip()->version();
-    const auto local_version = local_address->ip()->version();
-    if (remote_version != protocol_version || local_version != protocol_version) {
-      throw EnvoyException("failed to read proxy protocol");
+bool Filter::parseExtensions(int fd) {
+  while (proxy_protocol_header_.value().extensions_length_) {
```

Comment:
Could you add a comment that if in the future we ever parse the TLVs, we should only do it for `PROXY` and not `LOCAL`?  We want to skip over all of the local bytes.  We may or may not want to parse the TLVs for `PROXY`.

The spec only mentions TLVs for `PROXY`:

```
If the length specified in the PROXY protocol header indicates that additional
bytes are part of the header beyond the address information, a receiver may
choose to skip over and ignore those bytes, or attempt to interpret those
bytes.
```

For `LOCAL`, you just skip over them:

```
When a sender presents a
LOCAL connection, it should not present any address so it sets this field to
zero. Receivers MUST always consider this field to skip the appropriate number
of bytes and must not assume zero is presented for LOCAL connections.
```